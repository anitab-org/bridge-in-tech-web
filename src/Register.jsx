import React from "react";

import { withFormik, Form, Field, ErrorMessage } from "formik";
import * as Yup from "yup";


const RegisterForm = ({ values, errors, touched, isSubmitting }) => (
  <div className="container">
    <div className="row mb-5">
      <div className="col-lg-12 text-center">
        <h1 className="mt-5">Register Form</h1>
      </div>
    </div>
    <div className="row">
      <div className="col-lg-12">
        <Form>
            <div className="form-group">
                <label htmlFor="name">Name</label>
                <Field
                type="text"
                name="name"
                placeholder="Enter full name"
                className={`form-control ${
                    touched.name && errors.name ? "is-invalid" : ""
                }`}
                />
                <ErrorMessage
                component="div"
                name="name"
                className="invalid-feedback"
                />
            </div>
            <div className="form-group">
                <label htmlFor="username">Username</label>
                <Field
                type="text"
                name="username"
                placeholder="Enter a username"
                className={`form-control ${
                    touched.username && errors.username ? "is-invalid" : ""
                }`}
                />
                <ErrorMessage
                component="div"
                name="username"
                className="invalid-feedback"
                />
            </div>
            <div className="form-group">
                <label htmlFor="email">Email</label>
                <Field
                type="email"
                name="email"
                placeholder="Enter email"
                className={`form-control ${
                    touched.email && errors.email ? "is-invalid" : ""
                }`}
                />
                <ErrorMessage
                component="div"
                name="email"
                className="invalid-feedback"
                />
            </div>
            <div className="form-group">
                <label htmlFor="password">Password</label>
                <Field
                type={values.togglePassword ? "text" : "password"}
                name="password"
                placeholder="Enter password"
                className={`form-control ${
                    touched.password && errors.password ? "is-invalid" : ""
                }`}
                />
                <ErrorMessage
                component="div"
                name="password"
                className="invalid-feedback"
                />
            </div>
            <label>
                <Field
                type="checkbox"
                name="togglePassword"
                checked={values.togglePassword}
                />
                Show Password
            </label>
            <br></br>
            <br></br>
            <div className="form-group">
                <label htmlFor="passwordConfirm">Confirm Password</label>
                <Field
                type={values.togglePassword ? "text" : "password"}
                name="passwordConfirm"
                placeholder="Confirm password"
                className={`form-control ${
                    touched.passwordConfirm && errors.passwordConfirm
                    ? "is-invalid"
                    : ""
                }`}
                />
                <ErrorMessage
                component="div"
                name="passwordConfirm"
                className="invalid-feedback"
                />
            </div>
            <div>
                <a className="btn btn-danger" id="cancelbtn" href="/" role="button">
                Cancel
                </a>
                <button
                className="btn btn-success"
                type="submit"
                name="submit"
                value="Signup"
                disabled={isSubmitting}
                >
                {isSubmitting ? "Please wait..." : "Submit"}
                </button>
            </div>
        </Form>
      </div>
    </div>
  </div>
);

const RegisterFormik = withFormik({
  mapPropsToValues({ name, username, email, password, passwordConfirm, togglePassword }) {
    return {
        name: name || "",
        username: username || "",
        email: email || "",
        password: password || "",
        passwordConfirm: passwordConfirm || "",
        togglePassword: togglePassword || false
    };
  },
  validationSchema: Yup.object().shape({
    name: Yup.string()
        .required("name is required")
        .min(2, "Name must be 2 characters at minimum")
        .max(30, "Name must be 30 characters at maximum")
        .matches(
            /^[a-zA-Z/\s/-]+$/,
            "Must only contains alphabet but may have whtespace and dash '-' " 
        ),
    username: Yup.string()
        .required("Username is required")
        .min(5, "Username must be 5 characters at minimum")
        .max(25, "Username must be 25 characters at maximum")
        .matches(
            /^[a-zA-Z0-9/_]+$/,
            "May contain alphabets or numbers or undescore '_' "
        ),
    email: Yup.string()
      .email("Invalid email address")
      .required("Email is required")
      .matches(
          /^[a-zA-Z0-9/_/./+/-]+@+[a-zA-Z0-9/-]+\.[a-zA-Z0-9/-/.]+$/,
          "May contain alphabets, numbers, characters '_ . + -' and proper email format 'xxx@xxx.com'"
      ),
    password: Yup.string()
      .min(8, "Password must be 8 characters at minimum")
      .max(64, "Password must be 64 characters at maximum")
      .required("Password is required"),
    passwordConfirm: Yup.string()
      .oneOf([Yup.ref("password"), null], "Passwords must match")
      .required("Confirm Password is required")
  }),
  handleSubmit(values, { resetForm, setErrors, setSubmitting }) {
    setTimeout(() => {
      if (values.email === "test@test.com") {
        setErrors({ email: "User already exist" });
      } else {
        console.log(values);
        resetForm();
      }
      setSubmitting(false);
    }, 100000); // 
  }
})(RegisterForm);

export default function Register() {
    return <RegisterFormik />;
}