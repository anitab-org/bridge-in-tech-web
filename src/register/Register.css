.container {
    background: #F7F9FA;
    height: auto;
    width: 70%;
    margin: 5em auto;
    -webkit-box-shadow: 5px 5px 5px 0px rgba(0, 0, 0, 0.4);
    -moz-box-shadow: 5px 5px 5px 0px rgba(0, 0, 0, 0.4);
    box-shadow: 5px 5px 5px 0px rgba(0, 0, 0, 0.4);
}



label {
    color: #24B9B6;
    font-size: 1.2em;
    font-weight: 400;
}

h1 {
    color: #24B9B6;
    padding-top: .5em;
}


.error {
    color: red;
}
.error-message {
    color: #FF6565;
    padding: .5em .2em;
    height: 1em;
    position: absolute;
    font-size: .8em;
}

.field {
    border: 1px solid #1C6EA4;
    width: 100%;
    height: 2em;
}

.button-group {
    width: 100%;
}


.register-form {
width: 80%;
text-align: left;
padding-top: 2em;
padding-bottom: 2em;
}

.space-btw{
    width: 40%;
}

@media only screen and (max-width: 600px) {
    .space-btw{
        width: 20%;
    }
  }