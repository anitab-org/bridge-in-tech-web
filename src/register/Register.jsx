import React, { useState } from "react";
import "./Register.css";
import {BASE_API} from "../config";
import {SERVICE_UNAVAILABLE_ERROR} from "../messages";


export default function Register() {
    const [isValidName, setIsValidName] = useState(true);
    const [isValidUsername, setIsValidUsername] = useState(true);
    const [isValidEmail, setIsValidEmail] = useState(true);
    const [isValidPassword, setIsValidPassword] = useState(true);
    const [isValidConfirmPassword, setIsValidConfirmPassword] = useState(true);
    const [responseMessage, setResponseMessage] = useState(null);
    const [password, setPassword] = useState("")
    const [confirmPassword, setConfirmPassword] = useState("")
    
    const handleSubmit = async e => {
        e.preventDefault();
        if (confirmPassword === password){
            let payload = {
                need_mentoring: false,
                available_to_mentor: false
              }
              new FormData(e.target).forEach((value, key) => {
                if (key === "terms_and_conditions_checked" || key === "need_mentoring" || key === "available_to_mentor")
                  value = (value === "true") ? true : false   
                if (key !== "confirmPassword") payload[key] = value;
              });
            const requestRegister = {
                method: "POST",
                headers: {
                    "Accept": "application/json",
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(payload)
            };

            fetch(`${BASE_API}/register`, requestRegister)
                .then(async response => {
                    let data = await response.json();
                    if (response.status_code === 201)
                        return setResponseMessage(data.message);
                    setResponseMessage(data.message);
                })
                .catch(() => {
                    setResponseMessage(SERVICE_UNAVAILABLE_ERROR);
                });
        }
    }

    const validateName = e => {
        setIsValidName(e.target.checkValidity());
    };
    const validateUsername = e => {
        setIsValidUsername(e.target.checkValidity());
    };
    const validateEmail = e => {
        setIsValidEmail(e.target.checkValidity());
    };
    const validatePassword = e => {
        setPassword(e.target.value)
        setIsValidPassword(e.target.checkValidity());
    };
    const validateConfirmPassword = e => {
        setConfirmPassword(e.target.value)
        setIsValidConfirmPassword(e.target.checkValidity() && e.target.value === password);
    };

    return (
        <div className="container">
            <div className="row mb-5">
                <div className="col-lg-12 text-center">
                    <h1 className="mt-5">Register</h1>
                </div>
            </div>
            <div className="row">
                <div className="col-lg-12">
                    <form className="register-form mx-auto" onSubmit={handleSubmit}>
                        <form-group controlId="formName">
                            <p className="input-control">
                                <label htmlFor="name">Name :</label>
                                <input className="field"
                                    type="text"
                                    name="name"
                                    placeholder="Full Name"
                                    minLength={2}
                                    maxLength={30}
                                    pattern="^[a-zA-Z\s\-]+$"
                                    onChange={validateName}
                                    required
                                />
                            </p>
                            {!isValidName && (
                                <span className="error">Must be between 2-30 characters long. Can only contain alphabets, whitespace and dash '-'</span>
                            )}
                        </form-group>
                        <div><br></br></div>
                        <form-group controlId="formUserame">
                            <p className="input-control">
                                <label htmlFor="username">Username :</label>
                                <input className="field"
                                    type="text"
                                    name="username"
                                    placeholder="Username"
                                    minLength={5}
                                    maxLength={25}
                                    pattern="^[a-zA-Z0-9/_]+$"
                                    onChange={validateUsername}
                                    required
                                />
                            </p>
                            {!isValidUsername && (
                                <span className="error">Must be between 5-25 characters long. Can only contain alphabets, numbers and underscore '_'</span>
                            )}
                        </form-group>
                        <div><br></br></div>
                        <form-group controlId="formEmail">
                            <p className="input-control">
                                <label htmlFor="email">Email :</label>
                                <input className="field"
                                    type="email"
                                    name="email"
                                    placeholder="Email"
                                    pattern="^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$"
                                    onChange={validateEmail}
                                    required
                                />
                            </p>
                            {!isValidEmail && (
                                <span className="error">Must match standard email format xxx@xxx.xxx</span>
                            )}
                        </form-group>
                        <div><br></br></div>
                        <form-group controlId="formPassword">
                            <p className="input-control">
                                <label htmlFor="password">Password :</label>
                                <input className="field"
                                    type="password"
                                    name="password"
                                    placeholder="Password"
                                    minLength={8}
                                    maxLength={64}
                                    onChange={validatePassword}
                                    required
                                />
                            </p>
                            {!isValidPassword && (
                                <span className="error">Must be between 8-64 characters</span>
                            )}
                        </form-group>
                        <div><br></br></div>
                        <form-group controlId="formPassword">
                            <p className="input-control">
                                <label htmlFor="password">Confirm Password :</label>
                                <input className="field"
                                    type="password"
                                    name="confirmPassword"
                                    placeholder="Confirm Password"
                                    minLength={8}
                                    maxLength={64}
                                    onChange={validateConfirmPassword}
                                    required
                                />
                            </p>
                            {!isValidConfirmPassword && (
                                <span className="error">Must be same as password</span>
                            )}
                        </form-group>
                        <div><br></br></div>
                        <div><br></br></div>
                        <form-group>
                            <div className="row">
                                <div className="col-sm text-center">
                                    <label htmlFor="availability">Available to be a :</label>
                                    <form-group>
                                            <div className="mb-3">
                                                <div className="row">
                                                    <div className="col-sm">
                                                        <input
                                                            name="available_to_mentor"
                                                            aria-label="mentor"
                                                            type="checkbox"
                                                            value={"on" ? true : false}
                                                        />
                                                        <label htmlFor="mentor">Mentor</label>
                                                    </div>
                                                    <div className="col-sm">
                                                        <input
                                                            name="need_mentoring"
                                                            aria-label="mentee"
                                                            type="checkbox"
                                                            value={"on" ? true : false}
                                                        />
                                                        <label htmlFor="mentee">Mentee</label>
                                                    </div>
                                                </div>
                                            </div>
                                    </form-group>
                                </div>
                            </div>
                        </form-group>
                        <form-group controlId="formTermsCheck">
                            <div className="row">
                                <div className="col-sm">
                                    <p className="input-control">
                                        <input
                                            type="checkbox"
                                            name="terms_and_conditions_checked"
                                            aria-label="termsCheck"
                                            value={"on" ? true : false}
                                            required
                                        />
                                    </p>
                                </div>
                                <div className="col-sm-10">
                                    <label>
                                        By checking this box, I affirm that I have read and accept
                                        to be bound by the AnitaB.org Code of Conduct, Terms, and
                                        Privacy Policy. Further I consent to the use of my
                                        information for the stated purpose.
                                    </label>
                                </div>
                            </div>
                        </form-group>
                        <div>
                            {responseMessage && <span className="error" name="response" aria-label="response" role="alert">{responseMessage}</span>}
                        </div>
                        <div className="row">
                            <label>Already registered? Click on Login.</label>
                        </div>
                        <div className="row button-group">
                            <div className="col-sm">
                                <div className="row">
                                    <a className="btn btn-primary" id="loginbtn" href="/login" role="button">Login</a>
                                </div>
                            </div>
                            <div className="col-sm"></div>
                            <div className="col-sm">
                                <button className="btn btn-success"
                                    variant="success"
                                    type="submit"
                                    name="submit"
                                    value="Signup"
                                >Sign Up
                                </button>
                            </div>
                        </div>
                    </form>
                </div>
            </div>
        </div>
    );
}
