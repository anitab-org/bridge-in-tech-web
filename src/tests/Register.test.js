import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import '@testing-library/jest-dom';
import Register from "../register/Register";


test('renders Register', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<Register />);
    expect(renderer.getRenderOutput()).toMatchSnapshot()
});