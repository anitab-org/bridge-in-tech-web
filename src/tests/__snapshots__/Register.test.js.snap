// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Register 1`] = `
<div
  className="container"
>
  <div
    className="row mb-5"
  >
    <div
      className="col-lg-12 text-center"
    >
      <h1
        className="mt-5"
      >
        Register Form
      </h1>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-lg-12"
    >
      <form
        className="register-form mx-auto"
        onSubmit={[Function]}
      >
        <form-group
          controlId="formName"
        >
          <p
            className="input-control"
          >
            <label
              htmlFor="name"
            >
              Name :
            </label>
            <input
              className="field"
              maxLength={30}
              minLength={2}
              name="name"
              onChange={[Function]}
              pattern="^[a-zA-Z\\\\s\\\\-]+$"
              placeholder="Full Name"
              required={true}
              type="text"
            />
          </p>
        </form-group>
        <div>
          <br />
        </div>
        <form-group
          controlId="formUserame"
        >
          <p
            className="input-control"
          >
            <label
              htmlFor="username"
            >
              Username :
            </label>
            <input
              className="field"
              maxLength={25}
              minLength={5}
              name="username"
              onChange={[Function]}
              pattern="^[a-zA-Z0-9/_]+$"
              placeholder="Username"
              required={true}
              type="text"
            />
          </p>
        </form-group>
        <div>
          <br />
        </div>
        <form-group
          controlId="formEmail"
        >
          <p
            className="input-control"
          >
            <label
              htmlFor="email"
            >
              Email :
            </label>
            <input
              className="field"
              name="email"
              onChange={[Function]}
              pattern="^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\\\.[a-zA-Z0-9-.]+$"
              placeholder="Email"
              required={true}
              type="email"
            />
          </p>
        </form-group>
        <div>
          <br />
        </div>
        <form-group
          controlId="formPassword"
        >
          <p
            className="input-control"
          >
            <label
              htmlFor="password"
            >
              Password :
            </label>
            <input
              className="field"
              maxLength={64}
              minLength={8}
              name="password"
              onChange={[Function]}
              placeholder="Password"
              required={true}
              type="password"
            />
          </p>
        </form-group>
        <div>
          <br />
        </div>
        <div>
          <br />
        </div>
        <form-group>
          <div
            className="row"
          >
            <div
              className="col-sm text-center"
            >
              <label
                htmlFor="availability"
              >
                Available to be a :
              </label>
              <form-group>
                <div
                  className="mb-3"
                >
                  <div
                    className="row"
                  >
                    <div
                      className="col-sm"
                    >
                      <input
                        id="inline-checkbox-1"
                        name="available_to_mentor"
                        onChange={[Function]}
                        type="checkbox"
                        value={true}
                      />
                      <label
                        htmlFor="mentor"
                      >
                        Mentor
                      </label>
                    </div>
                    <div
                      className="col-sm"
                    >
                      <input
                        id="inline-checkbox-2"
                        name="need_mentoring"
                        onChange={[Function]}
                        type="checkbox"
                        value={true}
                      />
                      <label
                        htmlFor="mentee"
                      >
                        Mentee
                      </label>
                    </div>
                  </div>
                </div>
              </form-group>
            </div>
          </div>
        </form-group>
        <form-group
          controlId="formTermsCheck"
        >
          <div
            className="row"
          >
            <div
              className="col-sm"
            >
              <p
                className="input-control"
              >
                <input
                  name="terms_and_conditions_checked"
                  onChange={[Function]}
                  required={true}
                  type="checkbox"
                  value={true}
                />
              </p>
            </div>
            <div
              className="col-sm-10"
            >
              <label>
                By checking this box, I affirm that I have read and accept to be bound by the AnitaB.org Code of Conduct, Terms, and Privacy Policy. Further I consent to the use of my information for the stated purpose.
              </label>
            </div>
          </div>
        </form-group>
        <div />
        <div
          className="row"
        >
          <label>
            Already register? Login here.
          </label>
        </div>
        <div
          className="row button-group"
        >
          <div
            className="col-sm"
          >
            <div
              className="row"
            >
              <a
                className="btn btn-primary"
                href="/login"
                id="loginbtn"
                role="button"
              >
                Login
              </a>
            </div>
          </div>
          <div
            className="col-sm"
          />
          <div
            className="col-sm"
          >
            <button
              className="btn btn-success"
              name="submit"
              type="submit"
              value="Signup"
              variant="success"
            >
              Sign Up
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
